File: Sources/AislopdeskVideoProtocol/Geometry.swift
import Foundation

#if canImport(CoreGraphics)
import CoreGraphics
#endif

/// A pure 2-D point (host-space, points). Mirrors `CGPoint` but carries no platform
/// dependency, so `AislopdeskVideoProtocol` stays a leaf that compiles for macOS + iOS
/// and is unit-testable in isolation. Bridges to/from `CGPoint` where CoreGraphics
/// is available.
public struct VideoPoint: Equatable, Sendable {
    public var x: Double
    public var y: Double
    public init(x: Double, y: Double) { self.x = x
        self.y = y
    }
}

/// A pure 2-D size (points).
public struct VideoSize: Equatable, Sendable {
    public var width: Double
    public var height: Double
    public init(width: Double, height: Double) { self.width = width
        self.height = height
    }
}

/// A pure rectangle (origin + size), in whatever coordinate space the caller states.
public struct VideoRect: Equatable, Sendable {
    public var origin: VideoPoint
    public var size: VideoSize
    public init(origin: VideoPoint, size: VideoSize) { self.origin = origin
        self.size = size
    }

    public init(x: Double, y: Double, width: Double, height: Double) {
        origin = VideoPoint(x: x, y: y)
        size = VideoSize(width: width, height: height)
    }

    public var minX: Double { origin.x }
    public var minY: Double { origin.y }
    public var maxX: Double { origin.x + size.width }
    public var maxY: Double { origin.y + size.height }

    /// The area of intersection with `other` (0 when disjoint). Used by the
    /// multi-monitor coordinate-mapping screen pick.
    public func intersectionArea(_ other: Self) -> Double {
        let ix = max(0, min(maxX, other.maxX) - max(minX, other.minX))
        let iy = max(0, min(maxY, other.maxY) - max(minY, other.minY))
        return ix * iy
    }
}

/// How the decoded video is scaled into the on-screen layer (doc 17 §3.7). BOTH modes
/// PRESERVE the native aspect ratio — neither stretches/distorts:
/// - `.fit` (default) letterboxes/pillarboxes: the WHOLE remote window is visible, with
///   black bars on the longer axis when the pane's aspect differs.
/// - `.fill` covers the pane: NO bars, the video is scaled up until it covers the whole
///   pane and the overflowing axis is cropped by the viewport.
/// The user toggles between them ("nút fill"); `zoom`/`pan` then navigate within either
/// (e.g. pan to reach the cropped edges in `.fill`, or zoom in to read in `.fit`).
public enum VideoContentMode: Sendable, Equatable {
    case fit
    case fill
}

/// Aspect geometry — the **single source of truth** for where the decoded video is
/// actually drawn inside the layer (doc 17 §3.7). The Metal renderer scales the frame
/// (letterbox in `.fit`, cover-crop in `.fill`) so the video occupies a centred rect of
/// the layer; both the renderer (`fit` quad scale) and the input/cursor mapping derive
/// their geometry from this one function so render-forward and input-inverse can never
/// drift — including across a fit↔fill toggle.
///
/// Pure + platform-free (lives in the protocol leaf) so it is unit-testable in isolation
/// and usable from both `AislopdeskVideoClient` (renderer + input encoder + cursor compositor).
public enum AspectFit {
    /// The rect (origin + size) the displayed video occupies inside a `viewSize` layer,
    /// preserving the video's native aspect ratio. In `.fit` the rect is CONTAINED in the
    /// view (centred, with letterbox/pillarbox bars). In `.fill` the rect COVERS the view
    /// (centred, can exceed the view → negative origin / size > view; that overflow is the
    /// crop). Either way the rect is the exact region the renderer maps the full texture
    /// onto, and the region `normalize` inverts — so they always agree.
    ///
    /// MUST match `MetalVideoRenderer`'s `fit`-branch exactly: the renderer computes the
    /// same ratios in PIXELS (drawableSize × video pixel size), but aspect ratio is
    /// scale-invariant, so the rect is identical whether measured in points or pixels.
    ///
    /// - Parameters:
    ///   - viewSize: the layer's size (points, or pixels — scale-invariant).
    ///   - videoNativeSize: the decoded video's native size (same unit family).
    ///   - mode: `.fit` (contain, letterbox) or `.fill` (cover, crop). Default `.fit`.
    /// - Returns: the centred displayed-video rect. Falls back to the full `viewSize`
    ///   rect for any non-positive dimension (degenerate input is placed sensibly).
    public static func displayedVideoRect(
        viewSize: VideoSize,
        videoNativeSize: VideoSize,
        mode: VideoContentMode = .fit,
    ) -> VideoRect {
        let vw = videoNativeSize.width, vh = videoNativeSize.height
        let viewW = viewSize.width, viewH = viewSize.height
        guard vw > 0, vh > 0, viewW > 0, viewH > 0 else {
            return VideoRect(x: 0, y: 0, width: max(0, viewW), height: max(0, viewH))
        }
        // `.fit` scales to the SMALLER axis ratio (contain → the whole video sits inside,
        // bars on the longer axis). `.fill` scales to the LARGER axis ratio (cover → the
        // video fills the view, the longer axis overflows and is cropped). Both use a single
        // uniform `scale`, so neither distorts the aspect.
        let scaleX = viewW / vw, scaleY = viewH / vh
        let scale = (mode == .fit) ? min(scaleX, scaleY) : max(scaleX, scaleY)
        let w = vw * scale, h = vh * scale
        let ox = (viewW - w) / 2
        let oy = (viewH - h) / 2
        return VideoRect(x: ox, y: oy, width: w, height: h)
    }

    /// FORWARD render transform: maps a host-window-space point (points) to where it is
    /// drawn in the layer's view space (points). This is the exact inverse of
    /// ``AislopdeskVideoClient/InputEventEncoder/normalize(viewPoint:layerSize:videoNativeSize:zoom:pan:)``
    /// and the renderer's aspect-fit + zoom/pan crop, used to place the local cursor
    /// overlay where clicks actually land (doc 17 §3.3 / §3.7).
    ///
    /// 1. host point → source 0..1 (`hostPoint / videoNativeSize`).
    /// 2. invert the renderer's crop (`uv = (in.uv-0.5)·invZoom + 0.5 + pan`):
    ///    `displayUV = (sourceUV - 0.5 - pan)·zoom + 0.5`.
    /// 3. displayUV → view point inside the aspect-fit displayed rect.
    /// Pan is clamped identically to the renderer (`panLimit = 0.5·(1-invZoom)`).
    public static func viewPoint(
        forHostPoint hostPoint: VideoPoint,
        viewSize: VideoSize,
        videoNativeSize: VideoSize,
        zoom: Double = 1,
        pan: VideoPoint = VideoPoint(x: 0, y: 0),
        mode: VideoContentMode = .fit,
    ) -> VideoPoint {
        let su = videoNativeSize.width > 0 ? hostPoint.x / videoNativeSize.width : 0
        let sv = videoNativeSize.height > 0 ? hostPoint.y / videoNativeSize.height : 0
        let z = max(1, zoom)
        let invZoom = 1 / z
        let panLimit = 0.5 * (1 - invZoom)
        let px = min(max(pan.x, -panLimit), panLimit)
        let py = min(max(pan.y, -panLimit), panLimit)
        let du = (su - 0.5 - px) * z + 0.5
        let dv = (sv - 0.5 - py) * z + 0.5
        let r = displayedVideoRect(viewSize: viewSize, videoNativeSize: videoNativeSize, mode: mode)
        return VideoPoint(x: r.origin.x + du * r.size.width, y: r.origin.y + dv * r.size.height)
    }
}

#if canImport(CoreGraphics)
public extension VideoPoint {
    init(_ p: CGPoint) { self.init(x: Double(p.x), y: Double(p.y)) }
    var cgPoint: CGPoint { CGPoint(x: x, y: y) }
}

public extension VideoSize {
    init(_ s: CGSize) { self.init(width: Double(s.width), height: Double(s.height)) }
    var cgSize: CGSize { CGSize(width: width, height: height) }
}

public extension VideoRect {
    init(_ r: CGRect) {
        self.init(
            x: Double(r.origin.x),
            y: Double(r.origin.y),
            width: Double(r.size.width),
            height: Double(r.size.height),
        )
    }

    var cgRect: CGRect { CGRect(x: minX, y: minY, width: size.width, height: size.height) }
}
#endif

File: rust/aislopdesk-core/src/geometry.rs
//! Pure 2-D geometry + aspect-fit math — a port of Swift `Geometry.swift`.
//!
//! These types mirror `CGPoint`/`CGSize`/`CGRect` without any platform dependency so
//! the renderer's forward transform and the input encoder's inverse transform derive
//! from one shared source and can never drift (incl. across a fit↔fill toggle).

/// A pure 2-D point (host-space, points).
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct VideoPoint {
    /// Horizontal coordinate.
    pub x: f64,
    /// Vertical coordinate.
    pub y: f64,
}

impl VideoPoint {
    /// Builds a point.
    #[must_use]
    pub const fn new(x: f64, y: f64) -> Self {
        Self { x, y }
    }
}

/// A pure 2-D size (points).
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct VideoSize {
    /// Width.
    pub width: f64,
    /// Height.
    pub height: f64,
}

impl VideoSize {
    /// Builds a size.
    #[must_use]
    pub const fn new(width: f64, height: f64) -> Self {
        Self { width, height }
    }
}

/// A pure rectangle (origin + size), in whatever coordinate space the caller states.
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct VideoRect {
    /// Lower/upper-left origin (coordinate-space dependent).
    pub origin: VideoPoint,
    /// Extent.
    pub size: VideoSize,
}

impl VideoRect {
    /// Builds a rect from an origin and size.
    #[must_use]
    pub const fn new(origin: VideoPoint, size: VideoSize) -> Self {
        Self { origin, size }
    }

    /// Builds a rect from scalar components.
    #[must_use]
    pub const fn xywh(x: f64, y: f64, width: f64, height: f64) -> Self {
        Self {
            origin: VideoPoint::new(x, y),
            size: VideoSize::new(width, height),
        }
    }

    /// Minimum x (origin.x).
    #[must_use]
    pub const fn min_x(&self) -> f64 {
        self.origin.x
    }

    /// Minimum y (origin.y).
    #[must_use]
    pub const fn min_y(&self) -> f64 {
        self.origin.y
    }

    /// Maximum x (origin.x + width).
    #[must_use]
    pub fn max_x(&self) -> f64 {
        self.origin.x + self.size.width
    }

    /// Maximum y (origin.y + height).
    #[must_use]
    pub fn max_y(&self) -> f64 {
        self.origin.y + self.size.height
    }

    /// The area of intersection with `other` (0 when disjoint). Used by the
    /// multi-monitor coordinate-mapping screen pick.
    #[must_use]
    pub fn intersection_area(&self, other: &Self) -> f64 {
        let ix = 0.0_f64.max(self.max_x().min(other.max_x()) - self.min_x().max(other.min_x()));
        let iy = 0.0_f64.max(self.max_y().min(other.max_y()) - self.min_y().max(other.min_y()));
        ix * iy
    }
}

/// How the decoded video is scaled into the on-screen layer.
///
/// Both modes PRESERVE the
/// native aspect ratio (neither stretches): `Fit` letterboxes/pillarboxes so the whole
/// remote window is visible; `Fill` covers the pane, cropping the overflowing axis.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum VideoContentMode {
    /// Contain: the whole video sits inside the view (letterbox/pillarbox bars).
    Fit,
    /// Cover: the video fills the view; the longer axis overflows and is cropped.
    Fill,
}

/// Aspect-fit geometry — the single source of truth for where the decoded video is
/// drawn inside the layer. A port of Swift `AspectFit`.
pub mod aspect_fit {
    use super::{VideoContentMode, VideoPoint, VideoRect, VideoSize};

    /// The centred rect the displayed video occupies inside a `view_size` layer,
    /// preserving the native aspect ratio.
    ///
    /// In `Fit` the rect is contained (centred,
    /// with bars); in `Fill` it covers the view (centred, may exceed it — that overflow
    /// is the crop). Falls back to the full `view_size` rect for any non-positive
    /// dimension (degenerate input placed sensibly).
    #[must_use]
    pub fn displayed_video_rect(
        view_size: VideoSize,
        video_native_size: VideoSize,
        mode: VideoContentMode,
    ) -> VideoRect {
        let (vw, vh) = (video_native_size.width, video_native_size.height);
        let (cap_w, cap_h) = (view_size.width, view_size.height);
        if !(vw > 0.0 && vh > 0.0 && cap_w > 0.0 && cap_h > 0.0) {
            return VideoRect::xywh(0.0, 0.0, cap_w.max(0.0), cap_h.max(0.0));
        }
        // `Fit` scales to the SMALLER axis ratio (contain); `Fill` to the LARGER (cover).
        // A single uniform scale, so neither distorts the aspect.
        let scale_x = cap_w / vw;
        let scale_y = cap_h / vh;
        let scale = if mode == VideoContentMode::Fit {
            scale_x.min(scale_y)
        } else {
            scale_x.max(scale_y)
        };
        let w = vw * scale;
        let h = vh * scale;
        let ox = (cap_w - w) / 2.0;
        let oy = (cap_h - h) / 2.0;
        VideoRect::xywh(ox, oy, w, h)
    }

    /// FORWARD render transform: maps a host-window-space point to where it is drawn in
    /// the layer's view space.
    ///
    /// The exact inverse of the input encoder's `normalize` and
    /// the renderer's aspect-fit + zoom/pan crop. Pan is clamped identically to the
    /// renderer (`pan_limit = 0.5·(1 - 1/zoom)`).
    #[must_use]
    pub fn view_point(
        host_point: VideoPoint,
        view_size: VideoSize,
        video_native_size: VideoSize,
        zoom: f64,
        pan: VideoPoint,
        mode: VideoContentMode,
    ) -> VideoPoint {
        let su = if video_native_size.width > 0.0 {
            host_point.x / video_native_size.width
        } else {
            0.0
        };
        let sv = if video_native_size.height > 0.0 {
            host_point.y / video_native_size.height
        } else {
            0.0
        };
        let z = zoom.max(1.0);
        let inv_zoom = 1.0 / z;
        let pan_limit = 0.5 * (1.0 - inv_zoom);
        let px = pan.x.max(-pan_limit).min(pan_limit);
        let py = pan.y.max(-pan_limit).min(pan_limit);
        let du = (su - 0.5 - px) * z + 0.5;
        let dv = (sv - 0.5 - py) * z + 0.5;
        let r = displayed_video_rect(view_size, video_native_size, mode);
        VideoPoint::new(
            r.origin.x + du * r.size.width,
            r.origin.y + dv * r.size.height,
        )
    }
}

#[cfg(test)]
mod tests {
    use super::aspect_fit::{displayed_video_rect, view_point};
    use super::{VideoContentMode, VideoPoint, VideoRect, VideoSize};

    fn approx(a: f64, b: f64) {
        assert!((a - b).abs() < 1e-9, "{a} != {b}");
    }

    #[test]
    fn intersection_area_disjoint_is_zero() {
        let a = VideoRect::xywh(0.0, 0.0, 10.0, 10.0);
        let b = VideoRect::xywh(20.0, 20.0, 5.0, 5.0);
        approx(a.intersection_area(&b), 0.0);
    }

    #[test]
    fn intersection_area_overlap() {
        let a = VideoRect::xywh(0.0, 0.0, 10.0, 10.0);
        let b = VideoRect::xywh(5.0, 5.0, 10.0, 10.0);
        approx(a.intersection_area(&b), 25.0);
    }

    #[test]
    fn fit_letterboxes_wide_view() {
        // 16:9 video into a square view → pillarbox? video wider than view → letterbox top/bottom.
        let r = displayed_video_rect(
            VideoSize::new(100.0, 100.0),
            VideoSize::new(200.0, 100.0),
            VideoContentMode::Fit,
        );
        // scale = min(100/200, 100/100)=0.5 → 100x50 centred.
        approx(r.size.width, 100.0);
        approx(r.size.height, 50.0);
        approx(r.origin.x, 0.0);
        approx(r.origin.y, 25.0);
    }

    #[test]
    fn fill_covers_and_overflows() {
        let r = displayed_video_rect(
            VideoSize::new(100.0, 100.0),
            VideoSize::new(200.0, 100.0),
            VideoContentMode::Fill,
        );
        // scale = max(0.5,1.0)=1.0 → 200x100, origin x=-50.
        approx(r.size.width, 200.0);
        approx(r.size.height, 100.0);
        approx(r.origin.x, -50.0);
        approx(r.origin.y, 0.0);
    }

    #[test]
    fn degenerate_falls_back_to_view() {
        let r = displayed_video_rect(
            VideoSize::new(80.0, 60.0),
            VideoSize::new(0.0, 0.0),
            VideoContentMode::Fit,
        );
        assert_eq!(r, VideoRect::xywh(0.0, 0.0, 80.0, 60.0));
    }

    #[test]
    fn view_point_center_maps_to_rect_center() {
        let p = view_point(
            VideoPoint::new(100.0, 50.0), // center of a 200x100 video
            VideoSize::new(100.0, 100.0),
            VideoSize::new(200.0, 100.0),
            1.0,
            VideoPoint::new(0.0, 0.0),
            VideoContentMode::Fit,
        );
        // displayed rect is 100x50 at (0,25); center maps to (50, 50).
        approx(p.x, 50.0);
        approx(p.y, 50.0);
    }
}
